ENT-2978 change learner portal to support archiving and unarchiving courses (and re-label complete to archive)

binodpant opened the pull request:
This work handles UI changes as per: https://openedx.atlassian.net/browse/ENT-2978

It makes the following changes to the learner portal UI

* [X] : Renders an option 'Unarchive course' in the Gear icon dropdown, when markedDone=true, and course has not ended
* [X] : Sends a PATCH to the associated endpoint with markedDone:False when 
* [X] : Appropriately wires up necessary changes in the UI code (redux store, modal dialogs etc) associated with the Un archive workflow
* [X]: Conditionally shows a Continue learning button on archived courses, if course has not ended
* [x] : Process the field marked_done as boolean instead of string (this requires an api change as per PR https://openedx.atlassian.net/browse/ENT-3028)

Checklist to merge this:

- [X] : api changes merged https://github.com/edx/edx-enterprise/pull/792
- [x] : review feedback addressed
- [x] : tests added for new modal
- [X] : UI autoupdates after un archive (right now unarchive is not auto updating the UI, investigating). Created this task: https://openedx.atlassian.net/browse/ENT-3028
- [x] : This PR has been merged: https://github.com/edx/edx-enterprise/pull/803

Depends on: 
* https://github.com/edx/edx-enterprise/pull/792 : endpoint update
* https://openedx.atlassian.net/browse/ENT-3028 : fix for status in .endpoint. __UNTIL THIS FIX IS IN I HAVE A WORKAROUND INT HE UI CODE TO ALWAYS UPDATE THE COURSE STATUS TO IN_PROGRESS AFTER WE HAVE UNARCHIVED. THIS IS A HACK INDEED AND SHOULD BE REMOVED ONCE THE API ENDPOINT IS FIXED TO RETURN THE CORRECT COURSE STATUS FOR THE UNARCHIVE CASE__



### Reviewer notes etc:

Can send reviewers a screen record. I could not upload directly to the pr, and checking if we can use a service to link the screen record to the PR

binodpant commented:
Locally these changes show up as follows

### Main page
![Screen Shot 2020-06-08 at 4 39 29 PM](https://user-images.githubusercontent.com/528166/84078663-4629d280-a9a7-11ea-91dc-75936ce02e72.png)



### On clicking Archive this course
![Screen Shot 2020-06-08 at 4 40 39 PM](https://user-images.githubusercontent.com/528166/84078686-50e46780-a9a7-11ea-9f59-bb1db5510c9b.png)

adamstankiewicz commented on src/components/dashboard/main-content/course-enrollments/CourseSection.jsx:
```diff
@@ -28,6 +28,12 @@ class CourseSection extends React.Component {
+  getFormattedOptionalSubtitle = () => {
+    const { subtitle } = this.props;
+    if (!subtitle) { return null; }
+    return <p className="mt-2">{subtitle}</p>;
```
To clean up the spacing a bit more, might change the `className` for the subtitle to "mt-3 mb-0".

adamstankiewicz commented on src/components/dashboard/main-content/course-enrollments/course-cards/InProgressCourseCard.jsx:
```diff
@@ -66,7 +66,7 @@ const InProgressCourseCard = ({
         children: (
           <div role="menuitem">
-            Mark as complete
+            Archive this course
```
nit: "Archive course"

adamstankiewicz commented on src/components/dashboard/main-content/course-enrollments/course-cards/mark-complete-modal/MarkCompleteModal.jsx:
```diff
@@ -8,6 +8,9 @@ import MarkCompleteModalContext from './MarkCompleteModalContext';
 import ModalBody from './ModalBody';
 import { markCourseAsCompleteRequest } from './data/service';
 
+export const MARK_ARCHIVED_DEFAULT_LABEL = 'Archive this course';
```
nit: can remove the "this" from both of these labels.

adamstankiewicz commented on src/components/dashboard/main-content/course-enrollments/course-cards/mark-complete-modal/MarkCompleteModal.jsx:
```diff
@@ -71,13 +74,13 @@ const MarkCompleteModal = ({
     >
       <Modal
-        title="Mark course as complete"
+        title="Archive this course"
```
nit: can drop the "this"

adamstankiewicz commented on src/components/dashboard/main-content/course-enrollments/course-cards/mark-complete-modal/ModalBody.jsx:
```diff
@@ -13,11 +13,15 @@ const ModalBody = () => {
     <>
       {confirmError && <ModalError />}
       <p className="m-0">
-        Are you sure you want to mark
```
I think it makes sense from a design perspective to break this text out into 2 paragraphs, e.g.:

![image](https://user-images.githubusercontent.com/2828721/84165714-0106ae80-aa42-11ea-839b-0a86d52881ee.png)

binodpant commented on src/components/dashboard/main-content/course-enrollments/CourseEnrollments.jsx:
```diff
@@ -18,6 +18,12 @@ import {
 import * as actions from './data/actions';
 
+const ARCHIVED_COURSES_SECTION_SUBTITLE = `This section contains both the 
+  courses you have completed in the past and
```
hmm okay I had pretty much pasted what Joseph had posted in the ticket, wasn't sure do we need to run the changes by him? I could do sth like `See completed and voluntarily archived courses here` or similar

adamstankiewicz replied:
I primarily meant just within the code itself :) Like, lines 22-24 can probably be combined into one.

binodpant commented on src/components/dashboard/main-content/course-enrollments/CourseEnrollments.jsx:
```diff
@@ -134,7 +140,8 @@ export class CourseEnrollments extends Component {
-          title="Completed courses"
+          title="Archived courses"
+          subtitle={ARCHIVED_COURSES_SECTION_SUBTITLE}
           component={CompletedCourseCard}
           courseRuns={courseRuns.completed}
```
I totally agree. I think we should rename complete to archive going forward. In this PR I am accepting a little bit of duplication fo code also actually. My plan is to create a second PR which will include

* renaming complete to archive
* address some of the (intentional) duplication/DRY issues you will see in this PR. I thought it was better to have a new modal workflow since resuing the existing modal was leading to a lot of kung fu I would have to do in text names, constants, method names etc.

thoughts?

taliaedX replied:
The duplication you speak of - if you're referring to the ModalBody and ModalError and the like, I mean I imagine we could maybe refactor that to move out of the mark-complete-modal directory and have it take in props to indicate if the verbiage should reference completion or archiving.

But otherwise, I don't personally see it as a big deal, but I haven't dived into the front end aspects of these code bases too much, so I'm sure others may have stronger opinions.

adamstankiewicz replied:
Yeah, there's an opportunity to reduce duplication between the MarkCompleteModal and UnarchiveModal components. Not a high priority issue, though so probably doesn't make sense to include as part of this work.

binodpant replied:
resolving for this PR but I totally on board with some of the thoughts here.. I think we can cleanup some of the code for sure.. esp. around duplication between 'archive' and 'mark complete' course workflows as well as some language/naming cleanup. I jsut did not want to include a bigger change in this PR

binodpant commented:
So as per latest discussoin here is what we are going to do:

* have the gear icon show to the right always for all courses, even in archived section
* gear icon will have dropdown option to 'Undo archive' for manually archived courses (when marked_done = True)
* we will show a 'Continue learning' button for manually archived courses whose end date has not reached yet, but that will simply take to the LMS and will not unarchive the course. Un archiving will still need to be done via the Undo Archive dropdown option on the gear icon

binodpant commented:
@adamstankiewicz and I debugged that last issue
>  : UI autoupdates after un archvie (right now unarchive is not auto updating the UI, investigating)
and found that it was due ot the backend always returning 'complete' for the status of the enrollment even when we unarchive it (with marked_done=false)

I am going to dig into see if I can make it return the correct status, that may need a backend change. not sure yet.

binodpant commented:
Aha! I verified that if I hack the value of the status to 'in_progress' when I update the modifyCourseRunStatus() call in CompletedCourseCard.jsx then the Ui updates by moving the course back to in progress! 

```
const mapDispatchToProps = dispatch => ({
  modifyCourseRunStatus: (options) => {
    dispatch(updateCourseRunStatus({ ...options, status: 'in_progress' }));
  },
  modifyIsUnarchiveCourseCompleteStatus: (options) => {
    dispatch(updateIsUnarchiveCourseCompleteSuccess(options));
  },
});
```

So if we were to fix the backend to return 'in_progress' for this case for status, I think it will fix this particular issue.

binodpant commented:
All fixes in the PR now. All that is left to do is for me to work on some tests here to cover these changes, working on those. I think in this PR it makes sense to do a bit of testing but I did not want to add too granular because an upcoming refactor would then require a lot more changes.. If it is acceptable to do the 'higher' level testing the the UI updates in this PR, but not unit level tests, is that acceptable to reviewers? @taliaedX @adamstankiewicz

binodpant commented:
Damn! just noticed one more thing, this AC is not yet in place 
> Incomplete courses under the “Archived Courses” div provide a button to “Continue Learning”
 As long as this course run is still valid

Sorry missed that one, let me put that in there and I can put this up for final review

binodpant commented:
Added continue learning button to archived section and made it resuable and used it in both in progress and complete cards

binodpant commented on src/components/dashboard/main-content/course-enrollments/course-cards/mark-complete-modal/tests/MarkCompleteModal.test.jsx:
```diff
@@ -42,7 +42,7 @@ describe('<MarkCompleteModal />', () => {
       marked_done: 'True',
     });
-    expect(wrapper.find('.confirm-mark-complete-btn').hostNodes().text()).toEqual('Marking as complete...');
+    expect(wrapper.find('.confirm-mark-complete-btn').hostNodes().text()).toEqual(MARK_ARCHIVED_PENDING_LABEL);
```
indeed, will address the naming in a subsequent body of work, ty!

adamstankiewicz commented on src/components/dashboard/main-content/course-enrollments/course-cards/CompletedCourseCard.jsx:
```diff
@@ -1,36 +1,143 @@
+
 import { getAuthenticatedUser } from '@edx/frontend-platform/auth';
+import { sendTrackEvent } from '@edx/frontend-platform/analytics';
+import { courseEnded } from '../../../../../utils/common';
```
tiny nit: add a new line above `import { courseEnded }` so imports from installed packages are grouped together, and local imports are grouped.

also, random thought: perhaps `isCourseEnded` or `hasCourseEnded` to make it more clear its a boolean?

adamstankiewicz commented on src/components/dashboard/main-content/course-enrollments/course-cards/CompletedCourseCard.jsx:
```diff
@@ -1,36 +1,143 @@
+import { courseEnded } from '../../../../../utils/common';
+import ContinueLearningButton from './ContinueLearningButton';
+
+import {
```
tiny nit: swap these action imports with the `UnarchiveModal` import so that components are all grouped together, and then misc things like actions at the end of the imports.

adamstankiewicz commented on src/components/dashboard/main-content/course-enrollments/course-cards/CompletedCourseCard.jsx:
```diff
@@ -1,36 +1,143 @@
+
+  const [isModalOpen, setIsModalOpen] = useState(false);
+
+
```
tiny nit: extra newline

adamstankiewicz commented on src/components/dashboard/main-content/course-enrollments/course-cards/CompletedCourseCard.jsx:
```diff
@@ -1,36 +1,143 @@
+    modifyCourseRunStatus({
+      status: response.courseRunStatus,
+      courseId: response.courseRunId,
+      markedDone: response.markedDone.toLowerCase() === 'true',
```
Does the API return `markedDone` as a string? If so, wondering if we should fix the API to return an actual boolean so we don't need to convert the string to a boolean.

binodpant replied:
yup it does right "False" is what I got. So yes I think we should task out the fixing of api and subsequent changing of code here

binodpant replied:
tasked out to https://openedx.atlassian.net/browse/ENT-3047

adamstankiewicz commented on src/components/dashboard/main-content/course-enrollments/course-cards/CompletedCourseCard.jsx:
```diff
@@ -1,36 +1,143 @@
+  return (
+    <BaseCourseCard
+      buttons={renderButtons()}
+      dropdownMenuItems={getDropdownMenuItems({ markedDone })}
```
nit: is this an unnecessary argument? doesn't look like `getDropdownMenuItems` takes any arguments.

binodpant replied:
damn, you are right! I moved markdDone to the store already

adamstankiewicz commented on src/components/dashboard/main-content/course-enrollments/course-cards/CompletedCourseCard.jsx:
```diff
@@ -40,10 +147,25 @@ CompletedCourseCard.propTypes = {
-export default CompletedCourseCard;
+const mapDispatchToProps = dispatch => ({
+  modifyCourseRunStatus: (options) => {
+    // TODO remove this override of status once api is fixed
```
💯 thanks for adding a comment :)

adamstankiewicz commented on src/components/dashboard/main-content/course-enrollments/course-cards/InProgressCourseCard.jsx:
```diff
@@ -91,6 +86,7 @@ const InProgressCourseCard = ({
     modifyCourseRunStatus({
       status: response.courseRunStatus,
       courseId: response.courseRunId,
+      markedDone: response.markedDone.toLowerCase() === 'true',
```
similar comment as above -- should we update API to return a legit boolean so we don't have convert string to boolean?

binodpant replied:
absolutely agreed, shall I do this in two other subtasks, one for api and one for the ui? Perhaps it could also be a great starter ticket for a new team member, just having a thought moment here

adamstankiewicz commented on src/components/dashboard/main-content/course-enrollments/course-cards/mark-complete-modal/ModalBody.jsx:
```diff
@@ -13,11 +13,18 @@ const ModalBody = () => {
+        You will remain enrolled, but the course will
+        no longer appear as &quot;In Progress&quot;.
+      </p>
+      <p className="mt-1">
```
nit: perhaps `mt-2` to add a tiny bit more white space for separation between the paragraphs.

adamstankiewicz commented on src/components/dashboard/main-content/course-enrollments/course-cards/mark-complete-modal/ModalBody.jsx:
```diff
@@ -13,11 +13,18 @@ const ModalBody = () => {
+      </p>
+      <p className="mt-1">
+        As long as
+        your license is valid, you can resume the course
```
nit: Could this be condensed into a fewer number of lines (within the code, not changing the actual message).

adamstankiewicz commented on src/components/dashboard/main-content/course-enrollments/course-cards/unarchive-modal/ModalBody.jsx:
```diff
@@ -0,0 +1,31 @@
+        The course will show up in your &quot;In Progress&quot; section.
+      </p>
+      <p className="mt-1">
+        As long as
```
nit: similar as other comment, can this message be reduced to a fewer number of lines?

adamstankiewicz commented on src/components/dashboard/main-content/course-enrollments/course-cards/unarchive-modal/ModalError.jsx:
```diff
@@ -0,0 +1,33 @@
+            <FontAwesomeIcon className="mr-3" icon={faExclamationTriangle} />
+          </div>
+          <div>
+            Failure unarchiving
```
nit: Perhaps "An error occurred while unarchiving ..."?

binodpant commented:
Great review! Addressed your feedback, and also fixed an error in rendering ContinueLearningButton because of incorrect passing or destructured args...lol. Now rolling up sleeves for the tests!

taliaedX commented on src/components/dashboard/main-content/course-enrollments/course-cards/unarchive-modal/ModalBody.jsx:
```diff
@@ -13,16 +13,14 @@ const ModalBody = () => {
         {' '}
-        <a href={courseLink}>{courseTitle}</a>?
+        <a href={courseLink}>{courseTitle}</a>
         {' '}
```
What is `{' '}` doing btw? I can't tell if it's just adding a blank newline or somehow using variable replacement in a very obtuse way.

binodpant replied:
Great question! Yes it is basically the replacement for what we use d to as ```&nbsp;``` in older days. I think either nbsp or even `` based strings are also okay. I thikn this creates an extra text node, haven't validated on the browser...

adamstankiewicz replied:
Pretty sure it does create an extra text node using `{' '}`. Definitely feel free to update if it makes things cleaner.

taliaedX commented on src/components/dashboard/main-content/course-enrollments/tests/CourseEnrollments.test.jsx:
```diff
@@ -224,4 +227,20 @@ describe('<CourseEnrollments />', () => {
+    const enterpriseConfig = {
+      uuid: 'test-enterprise-uuid',
+    };
+    const wrapper = mount((
```
oh wow a proper mount. I've only ever used `shallow` 😂

binodpant replied:
yup I think with enzyme shallow was used a lot. I think RTL encourages (or even just defaults to full mounting and rendering... I have to dig deeper into render but I am pretty sure it does that). this FAQ answers some of what the RTL render() does and how to achieve shallow mount behavior if desired ...  https://testing-library.com/docs/react-testing-library/faq. doing full does require more work in mocking and handling side effects .. but it also give more realistic testing .. we will have to see case by case I guess ... :)

taliaedX commented on src/utils/common.js:
```diff
@@ -0,0 +1,8 @@
+import moment from 'moment';
+
+const isCourseEnded = (endDate) => moment(endDate) < moment();
```
how does moment differ from just using `Date.now`?

binodpant replied:
I think for this use case they are equivalent ... ```It's similar to native JavaScript's new Date().``` borrowed from some doc I read.. but I think moment in general makes working with date and time easier, handles locales better etc.. not too relevant to this pr I guess but yeah ...

binodpant commented:
Hi @taliaedX @adamstankiewicz I feel like if this test passes we should merge this change in even if the testing is not comprehensive. The reasn is thatt I want to identify and write more test cases for this from a more RTL style approach. Right now I am still just chekcing that a mock function is called but a real good way is to test that the ui responded correctly. I would want to do that but sinvce that will take quite a bit more time, I am inclined to move forward with this amount of testing 

subject to you guys' feedback of course

Thanks!

adamstankiewicz commented on src/components/dashboard/main-content/course-enrollments/tests/CourseEnrollments.test.jsx:
```diff
@@ -1,226 +1,90 @@
+  );
+}
+
+test('loads enrollments component', () => {
```
really like the syntax for React Testing Library 💯

binodpant replied:
indeed there is much to loave about it for sure, I love how Kent C. Dodds is vocal about issues enzyme has and how this tool is meant to address it. I find that many cases are not easy or possible to automated right now with this tool though since it no longer recommends using CSS classes or selectors as a way to find elements (it is possible though). Hence I think we have to go case by case and update ui elements if needed to have correct role + name combinations so we can mostly use stuff like `getbyRole('button', {name: <name>})` or any other field, which ensure that the fields used in tests are also accessible ... still much to learn and update here

adamstankiewicz commented on src/components/dashboard/main-content/course-enrollments/course-cards/CompletedCourseCard.jsx:
```diff
@@ -1,36 +1,142 @@
+  const getDropdownMenuItems = () => {
+    // Only courses that are manually archived (markedDone) should show a unarchive option.
+    // if course is ended or completed, you cannot unarchive it
+    // TODO: we also need to add || courseRunStatus === 'completed' once api returns correct status
```
Just curious about the sequencing of the API fix to return correct status and this PR. Are you hoping to get this merged in, fix the API to return correct status, and then come back and update this repo?

binodpant replied:
Yes, this part of the functionality does not actually work correctly until we fix the api issue. If I add this chekc now, it means we always hide the unarchive button. I would like to revisit this as soon as the api + this approach is finalized. I now think that using ENT-3028 as a pin to do that is a good idea, so I will leave that ticket around

binodpant replied:
to be clear, it does work correctly but it fails to hide this button for courses that are actually complete.

adamstankiewicz commented on src/components/dashboard/main-content/course-enrollments/tests/CourseEnrollments.test.jsx:
```diff
@@ -1,226 +1,90 @@
+
+const defaultInitialProps = defaultInitialEnrollmentProps({ genericMockFn });
+
+// TODO not sure why and if we need the sidebarComponent here
```
Back when the UX team gave us the original wireframes/mocks for this page, there was a requirement for small screens (aka mobile) to show the sidebar below the "My courses in progress" section but above the "Upcoming"/"Archived" sections (as opposed to having the sidebar rendered after all the course sections). To be honest, I'm not really sure what the rationale for that was but I just pinged Brandon C. (UX) to get clarification if this is still a requirement.

If it isn't, we could stop rendering the `Sidebar` within this `CourseEnrollments` section. I will follow-up with you once I hear back from Brandon C about this, but for now, yeah... `Sidebar` gets rendered within `CourseEnrollments` for mobile [between the in-progress and upcoming/archived sections](https://github.com/edx/frontend-app-learner-portal-enterprise/blob/master/src/components/dashboard/main-content/course-enrollments/CourseEnrollments.jsx#L124).

binodpant replied:
awesome! that is great context and help, we can and should do any actions as necessary after this merges and update the tests too thanks

binodpant commented:
Merging now, will discuss/address any raised concerns here in subsequent work, in the spirit of continuous improvement. We do need to handle the status correctly as part of ENT-3028 or related work post this though
